### PR TITLE
Update to Spark 2.4

### PR DESCRIPTION
**Pull request by chrisvittal**
I have no idea if I've updated everything I need to update here. I haven't pushed the updated CI
build image or changed cloudtools (yet).

Update: pushed ci build image.

cc @cseed @danking

### CODE REVIEW

File: hail/build.gradle [hunk 1]
```diff
@@ -200,13 +200,6 @@ dependencies {
     compile 'org.http4s:http4s-blaze-core_' + scalaMajorVersion + ':0.12.3'
     compile 'org.http4s:http4s-blaze-client_' + scalaMajorVersion + ':0.12.3'
     compile 'org.http4s:http4s-blaze-server_' + scalaMajorVersion + ':0.12.3'
-    compile 'org.json4s:json4s-core_' + scalaMajorVersion + ':3.2.10'
```
Comment (tpoterba) [inline hail/build.gradle]: we don't need this?

Comment (chrisvittal) [reply hail/build.gradle]: Apparently not. :man_shrugging:

File: hail/build.gradle [hunk 2]
```diff
@@ -44,7 +44,7 @@ String hailShortVersion = MAJOR.toString() + "." + MINOR.toString()
 String hailPipVersion = hailShortVersion + "." + PATCH + System.getProperty("hail.pip-version-suffix", "")
 String[] deployedSparkVersions = new File("deployed-spark-versions.txt").readLines()
 
-String sparkVersion = System.getProperty("spark.version", "2.2.0")
+String sparkVersion = System.getProperty("spark.version", "2.4.0")
```
Comment (tpoterba) [inline hail/build.gradle]: the fail message below should be 2.4

File: hail/build.gradle [hunk 3]
```diff
@@ -44,21 +44,21 @@ String hailShortVersion = MAJOR.toString() + "." + MINOR.toString()
...
 String scalaMajorVersion = '2.11'
 
-if (sparkVersion == '2.2.0') {
+if (sparkVersion == '2.4.0') {
     if (py4jVersion == null)
-      py4jVersion = '0.10.4'
+      py4jVersion = '0.10.7'
     if (breezeVersion == null)
-      breezeVersion = '0.13.1'
+      breezeVersion = '0.13.2'
```
Comment (danking) [inline hail/build.gradle]: we should keep the settings for old spark versions. I once went through and verified all the breezeVersions: https://github.com/hail-is/hail/pull/5130/files#diff-3c5fa39d01cc793629790d9a91cf9162L57

Basically 2.2.0 is `0.13.1` and all the other 2.x versions are `0.13.2`

File: hail/python/hail/docs/getting_started.rst [hunk 4]
```diff
@@ -17,8 +17,8 @@ Regardless of installation method, you will need:
   <https://www.anaconda.com/download/>`_
```
Comment (danking) [inline hail/python/hail/docs/getting_started.rst]: check `getting_started_developing.rst` too

File: hail/build.gradle (hunk 2 above)
Comment (danking) [reply hail/build.gradle]: Changing the default is a backwards incompatible change to our build system. Not sure how we feel about doing that to those that build on us.

File: hail/build.gradle (hunk 3 above)
Comment (tpoterba) [reply hail/build.gradle]: omg I spent 15 minutes looking for that yesterday across zulip/forum/email but didn't check the Make PR

**Comment by konradjk** [comment]
Yeah, is this going to deploy new jars only on 2.4 or also 2.2 for a while? This is good in the long run, but might be nice to keep 2.2 going in case there are any unintended consequences

**Comment by danking** [comment]
this will only do 2.4, but it's not much work to continue to provide both

**Comment by danking** [comment]
it looks like there's no source code changes

**Comment by tpoterba** [comment]
Dan, since you have a much better handle on how to get this in, can I give primary responsibility for this review to you?

I'm happy to accept a PR review in return.

**Comment by danking** [comment]
accepted

File: hail/build.gradle [hunk 5]
```diff
@@ -44,24 +44,30 @@ String hailShortVersion = MAJOR.toString() + "." + MINOR.toString()
...
-    if (breezeVersion == null)
-      breezeVersion = '0.13.1'
-} else {
-  if (breezeVersion == null || py4jVersion == null)
-    ant.fail('Unknown Spark version ' + sparkVersion + '. Set breeze.version and py4j.version properties for Spark ' + sparkVersion + '.')
+if (py4jVersion == null) {
+  if (sparkVersion == '2.2.0')
+    py4jVersion = '0.10.4'
+  else if (sparkVersion == '2.3.0')
+    py4jVersion = '0.10.6'
```
Comment (danking) [inline hail/build.gradle]: what about 2.2.1 and 2.3.1 and friends? Should we truncate to the first three characters of sparkVersion for these checks?

File: hail/python/hail/dev-environment-py3.7.yml [hunk 6]
```diff
@@ -14,9 +14,10 @@ dependencies:
 - pip=18.0
 - pytest=3.9.2
 - pytest-html=1.19
+- flask=0.12.2
```
Comment (danking) [inline hail/python/hail/dev-environment-py3.7.yml]: flask shouldn't be in the hail dev environment, we don't use it in hail.

We should probably get rid of these files anyway, we're just using the root requirements.txt now

File: hail/python/hail/docs/getting_started_developing.rst [hunk 7]
```diff
@@ -10,9 +10,9 @@ Requirements
 You'll need:
 
 - `Java 8 JDK <http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html>`_
-- `Spark 2.2.0 <https://www.apache.org/dyn/closer.lua/spark/spark-2.2.0/spark-2.2.0-bin-hadoop2.7.tgz>`_
+- `Spark 2.4.0 <https://www.apache.org/dyn/closer.lua/spark/spark-2.4.0/spark-2.4.0-bin-hadoop2.7.tgz>`_
 
-  - Hail will work with other bug fix versions of Spark 2.2.x, but it *will not* work with Spark 1.x.x, 2.0.x, or 2.1.x.
+  - Hail will work with other bug fix versions of Spark 2.4.x, but it *will not* work with Spark 1.x.x, 2.0.x, or 2.1.x.
```
Comment (danking) [inline hail/python/hail/docs/getting_started_developing.rst]: > Hail is compatible with Spark versions 2.2.x, 2.3.x, and 2.4.x.

File: hail/build.gradle (hunk 5 above)
Comment (chrisvittal) [reply hail/build.gradle]: These are the only ones that are different. Spark 2.2.1 is using 0.10.7, Spark 2.3.1 is using 0.10.7.

File: hail/python/hail/dev-environment-py3.7.yml (hunk 6 above)
Comment (chrisvittal) [reply hail/python/hail/dev-environment-py3.7.yml]: Will do

**Comment by danking** [comment]
ship it

**Comment by danking** [comment]
I added a cloudtools version pin in the cluster tests.

**Comment by chrisvittal** [comment]
The new cloudtools sha file needs to exist. I added it.

**Comment by danking** [comment]
Something is wrong with batch tests ...

**Comment by danking** [comment]
CI is quiet, so I'll retest now.

**Comment by danking** [comment]
I'll have to look into test_callback though.

**Comment by chrisvittal** [comment]
I believe that my work in #5813 will resolve the issues that we are seeing here. Please take a look.

**Pull request merged**